### ISSUE TITLE
My application used RunWithinApplication, but now it's deprecated. What should I do?

### ISSUE THREAD

**Issue by joozek78** [Starcounter/Home#149]
`RunWithinApplication` has been deprecated, but no guidance is given as to how to replace its usage. At least @malx122 is using it extensively. Can someone provide some migration instructions?

**Comment by alemoi** [comment]
I would say that it totally depends on the usage. The problem is that this concept totally violates the apps separation, so if we take it by concrete examples, then we solve it. 

One of such legit examples is trying to access all SQL tables.

**Comment by alemoi** [comment]
I didn't get any answer but would like to inform that `Db.GlobalSQL` is in place now. Its to be used to query all database classes using **fully namespaced** classes names. CC @miyconst @warpech @mtornwall @joozek78

**Comment by mmnosek** [comment]
@alemoi could you elaborate more how it works? What type of object is returned from such query? For me, it looks like anti-pattern and also violates apps separation rule. Shouldn't this kind of features be available in some other API than `Db`? It encourages developers to do really bad things 😭  but I understand that in some cases it's necessary.

**Comment by per-samuelsson** [comment]
> Shouldn't this kind of features be available in some other API than `Db`?

Most definitely. We should strive hard to keep `Db.*` as small as possible, as that is probably our broadest API in terms of usage. `Db.*` should preferably be self-explanatory and concerned only with things used by 99.9% of all apps.

A thing such as "global SQL" should not even be in the root Starcounter namespace.

**Comment by alemoi** [comment]
I agree regarding Db namespace and have moved the `Db.GlobalSQL` to `Db.Advanced.GlobalSQL`. Hope that is more convenient.

**Comment by alemoi** [comment]
> @alemoi could you elaborate more how it works? What type of object is returned from such query?

@mmnosek its no different from a standard `Db.SQL` except that it allows access to all database classes by fully namespaced name. The function documentation provides this info. Please let me know if this is not enough. The following nightly will have it in `Db.Advanced.GlobalSQL`

**Comment by alemoi** [comment]
Just a reminder: so far I didn't get any other justified use for `RunWithinApplication` except this one with SQL. Do you know any @joozek78, after talking with @malx122 maybe? Thanks!

**Comment by bigwad** [comment]
RunWithinApplication is vital to allow scenarios such as https://github.com/Starcounter/level1/issues/3210
That is to call Db.SQL from a scheduled task.

**Comment by bigwad** [comment]
@alemoi, I've put back usage of RunWithinApplication in [this code](https://github.com/Starcounter/level1/commit/956942886ad388cda52243987a2c5fca489a597f), until we have something for replacement.

**Comment by mmnosek** [comment]
@alemoi @bigwad  [your change](https://github.com/Starcounter/level1/commit/956942886ad388cda52243987a2c5fca489a597f) just increased benchmark numbers 5 times. I just want to make sure - without `RunWithinApplication` code is run on one thread, having this allows code to be executed on any available thread? 

Could you please explain why it's like this?

**Comment by bigwad** [comment]
@mmnosek, it's not related to RunWithinApplication - it's internals of tpcc test. I think previously it was overparallelized - I mean it executed in parallel some transaction it shouldn't. But I'll double check it.

**Comment by mmnosek** [comment]
Please do, it makes a huge difference, I would really like to understand it.

**Comment by bigwad** [comment]
@mmnosek, I checked tpcc and latest change might affect its performance in some cases, but I'm not sure you're talking about the same issue. 
What benchmark are you talking about?

**Comment by mmnosek** [comment]
I'm doing TPCC implemented in Level1: https://github.com/Starcounter/level1/tree/develop/src/Samples/TPCC

The test I do is just calling`/all_no_io` handler, so internal transactions without any I/O. Without your change above I had around 2000tps, after adding `RunWithinApplication` I got around 10000tps. So it looks like a massive difference.

**Comment by bigwad** [comment]
Ok, I see. 
The thing is - tpcc used to have an issue with error handling in /all_no_io handler you use. Actually, exceptions were just ignored. But there were no actual exceptions, so tpcc had run just fine until @alemoi removed RunWithinApplication call. Then /all_no_io started to fail, but nobody noticed that and I think that performance difference you see is just the fail path runs way slower. With my recent change I put back RunWithinApplication call and fixed error handling, hence reverted the test back to the success path.

PS.
Error handling I introduced had a side effect of waiting for the transaction log. I've changed that with https://github.com/Starcounter/level1/commit/a0dc6db9bdd7ca70bd3347088fcb57faca974631
